[c10d] Prototype of `group_split` for dist2 work

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157716

This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[WIP] Add prototype of group_split and group_merge

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split and group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split and group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split and group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split and group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split and group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1316,6 +1316,52 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
+  }
+  ncclOpts->split_from =
+      c10::intrusive_ptr<ProcessGroupNCCL>::unsafe_reclaim_from_nonowning(this);
+  ncclOpts->global_ranks_in_group = globalRanksInGroup;
```
This will be cleaned up but I need to send out a separate PR because it contains internal change so I want to separate from this one.

fduwjj replied:
Done in https://github.com/pytorch/pytorch/pull/157978 and will rebase this PR on top of it.

d4l3k reviewed (commented):
Looks reasonable to me and definitely in the right direction

fduwjj commented on torch/distributed/distributed_c10d.py:
```diff
@@ -129,6 +129,7 @@
     "reduce_scatter_tensor",
     "get_node_local_rank",
     "split_group",
+    "_world",
```
opps, let me remove it.. this is from previous WIP..

fduwjj commented on torch/distributed/_dist2.py:
```diff
@@ -134,3 +135,64 @@ def new_group(
     return _BACKENDS[backend](store, rank, world_size, timeout, device, pg_options)
+
+
+def split_group(
```
I see, so in that case, we only expose the API via pybind?

fduwjj commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -577,6 +577,27 @@ size_t hashTensors(const std::vector<at::Tensor>& tensors) {
+// list and ensure its range of 32-bit int.
+int genNcclSplitColor(const std::vector<int>& ranks) {
+  // Combine the hash values using a simple reducer (std::hash + fold)
+  std::size_t combined_hash = std::accumulate(
```
The original implementation use hash in python here: https://github.com/pytorch/pytorch/blob/b40c0b61eb5f5b6104cb732e709819e84518faa7/torch/distributed/distributed_c10d.py#L4934

If we use min(ranks) how do we different between [0, 1] and [0,1,2,3] if the original pg has a world_size of 8, for example?

d4l3k replied:
I think that configuration is strictly invalid.

What would it mean for one machine to have a PG with world_size 2 and another with world_size 4 with different members?

fduwjj replied:
I guess technically they are all valid sub-PG? And actually in Ads, they do have something like this...

fduwjj committed:
Update on "[WIP] Add prototype of group_split"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add prototype of group_split"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -695,6 +695,42 @@ const std::vector<uint64_t>& ProcessGroupGloo::groupRanks() const {
+  }
+  glooOpts->global_ranks_in_group = std::move(globalRanksInGroup);
+  // TODO: Figure out a better way for split backend name.
+  glooOpts->group_name = c10::str(pg_uid_, ":split");
```
somehow the base class does not have `group_name` attribute, unless we are ok to add one in the base class.

d4l3k replied:
doesn't hurt to add one

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -158,6 +159,63 @@ void ProcessGroup::release_resources() {
+        opts.has_value() ? opts.value() : parentBackend->getBackendOptions();
+    backendOpts->timeout =
+        timeout.has_value() ? timeout.value() : backendOpts->timeout;
+    auto splitBackend = parentBackend->splitBackend(sorted_ranks, backendOpts);
```
should we make backendOpts optional in splitBackend signature? Then we can get rid of `getBackendOptions`

fduwjj replied:
Hmm this is already in cpp, I don't see a big difference? I mean if we do move the timeout/groupName into backend, we are adding more args to `splitBackend` which I am not a big fan of.

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -158,6 +159,63 @@ void ProcessGroup::release_resources() {
+    std::string groupDesc = desc.has_value()
+        ? desc.value()
+        : c10::str(
+              getGroupDesc(), ":split:", parentBackend->getCommSplitCounter());
```
ah this is interesting -- makes sense but I don't quite love it

d4l3k commented on torch/csrc/distributed/c10d/Backend.hpp:
```diff
@@ -418,6 +436,13 @@ class TORCH_API Backend : public torch::CustomClassHolder {
         c10::str("Backend ", getBackendName(), " does not support getError"));
   }
 
+  virtual uint64_t getCommSplitCounter() const {
```
Does this need to differ on a backend basis or can we just have a simple impl in ProcessGroup. We could rename this just `incrementSplitCount()` and have it increment and return the value.

d4l3k commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2063,6 +2063,14 @@ communication mechanism.
+              py::arg("timeout") = std::nullopt,
+              py::arg("opts") = std::nullopt,
+              py::arg("groupDesc") = std::nullopt,
+              py::call_guard<py::gil_scoped_release>())
```
do we also need to add this to PyProcessGroup?

fduwjj replied:
oh I almost forgot about it... Sure, let me add it.

d4l3k reviewed (approved):
This is looking good! Left a few comments

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -i

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Prototype of `group_split` for dist2 work"


This is to implement group_split as proposed in [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]
